src/lib.rs

```diff
@@ -455,68 +597,33 @@ impl<T> Receiver<T> {
     /// Panics if called after this receiver has been polled asynchronously.
     #[cfg(feature = "std")]
     pub fn recv_ref(&self) -> Result<T, RecvError> {
-        let channel_ptr = self.channel_ptr;
-        let channel = unsafe { channel_ptr.as_ref() };
-
-        match channel.state.load(SeqCst) {
-            // The sender is alive but has not sent anything yet. We prepare to park.
-            EMPTY => {
-                // Conditionally add a delay here to help the tests trigger the edge cases where
-                // the sender manages to be dropped or send something before we are able to store
-                // our waker object in the channel.
-                #[cfg(oneshot_test_delay)]
-                std::thread::sleep(std::time::Duration::from_millis(10));
-
-                // Write our waker instance to the channel.
-                unsafe { channel.write_waker(ReceiverWaker::current_thread()) };
-
-                match channel
-                    .state
-                    .compare_exchange(EMPTY, RECEIVING, SeqCst, SeqCst)
-                {
-                    // We stored our waker, now we park until the sender has changed the state
-                    Ok(EMPTY) => loop {
-                        thread::park();
-                        match channel.state.load(SeqCst) {
-                            // The sender sent the message while we were parked.
-                            // We take the message and mark the channel disconnected.
-                            MESSAGE => {
-                                channel.state.store(DISCONNECTED, SeqCst);
-                                break Ok(unsafe { channel.take_message() });
-                            }
-                            // The sender was dropped while we were parked.
-                            DISCONNECTED => break Err(RecvError),
-                            // State did not change, spurious wakeup, park again.
-                            RECEIVING => (),
-                            _ => unreachable!(),
+        self.start_recv_ref(
+            || RecvError,
+            |channel| {
+                loop {
+                    thread::park();
+
+                    // ORDERING: we use acquire ordering to synchronize with the write of the message
+                    match channel.state.load(Acquire) {
+                        // The sender sent the message while we were parked.
+                        // We take the message and mark the channel disconnected.
+                        MESSAGE => {
+                            // ORDERING: the sender is inactive at this point so we don't need to make
+                            // and reads or writes visible to the sending thread
```

Comment:
"to make and reads" should probably be "to make our reads"?

Reply:
I think I meant "to make any reads"